File: app/src/main/java/com/example/shercofaqapp/viewmodel/BikeAdapter.java
package com.example.shercofaqapp.viewmodel;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.example.shercofaqapp.R;
import com.example.shercofaqapp.databinding.GarageItemBinding;
import com.example.shercofaqapp.model.Bike;

import java.util.ArrayList;

public class BikeAdapter extends RecyclerView.Adapter<BikeAdapter.BikeViewHolder> {

    private OnItemClickListener onItemClickListener;
    private ArrayList<Bike> bikesArrayList = new ArrayList<>();

    @NonNull
    @Override
    public BikeViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        GarageItemBinding garageItemBinding = DataBindingUtil.inflate(
                LayoutInflater.from(parent.getContext()), R.layout.garage_item,
                parent, false
        );

        return new BikeViewHolder(garageItemBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull BikeViewHolder holder, int position) {

        Bike bike = bikesArrayList.get(position);
        holder.garageItemBinding.setBike(bike);

    }

    @Override
    public int getItemCount() {
        return bikesArrayList.size();
    }

    class BikeViewHolder extends RecyclerView.ViewHolder {

        GarageItemBinding garageItemBinding;

        public BikeViewHolder(@NonNull GarageItemBinding garageItemBinding) {
            super(garageItemBinding.getRoot());

            this.garageItemBinding = garageItemBinding;
            garageItemBinding.getRoot().setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {

                    int position = getAdapterPosition();

                    if (onItemClickListener != null && position != RecyclerView.NO_POSITION) {
                        onItemClickListener.onItemClick(bikesArrayList.get(position));
                    }

                }
            });

        }
    }

    public interface OnItemClickListener {

        void onItemClick(Bike bike);

    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    public void setBikesArrayList(ArrayList<Bike> bikesArrayList) {
        this.bikesArrayList = bikesArrayList;
        notifyDataSetChanged();
    }

}

File: app/src/main/java/com/example/shercofaqapp/viewmodel/BikeAdapter.kt
package com.example.shercofaqapp.viewmodel

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.LiveData
import androidx.recyclerview.widget.RecyclerView
import com.example.shercofaqapp.R
import com.example.shercofaqapp.databinding.GarageItemBinding
import com.example.shercofaqapp.model.Bike

class BikeAdapter : RecyclerView.Adapter<BikeAdapter.ViewHolder>() {

    private var bikeList = ArrayList<Bike>()

    class ViewHolder(item: View) : RecyclerView.ViewHolder(item) {

        val binding = GarageItemBinding.bind(item)

        fun bind(bike: Bike) = with(binding) {

            bikeNameTextView.text = bike.bikeName
            bikeModelYearTextView.text = bike.bikeModelYear
            bikeEngineTypeTextView.text = bike.bikeEngineType
            bikeEngineVolumeTextView.text = bike.bikeEngineVolume
            bikeEditionTextView.text = bike.bikeEdition
            bikeImage.setImageResource(bike.bikeImage)

        }

    }

    // Create new views (invoked by the layout manager)
    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): ViewHolder {

        // Create a new view, which defines the UI of the list item
        return ViewHolder(item = LayoutInflater.from(viewGroup.context).inflate(R.layout.garage_item, viewGroup, false))

    }

    // Replace the contents of a view (invoked by the layout manager)
    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {

        viewHolder.bind(bikeList[position])

    }

    // Return the size of your dataset (invoked by the layout manager)
    override fun getItemCount(): Int {

        return bikeList.size

    }

    fun addBike(bike: Bike) {

        bikeList.add(bike)
        notifyDataSetChanged()

    }

    fun addBikeList(bikeList: ArrayList<Bike>) {

        this.bikeList = bikeList
        notifyDataSetChanged()

    }

}